compiler/rustc_middle/src/ty/util.rs

```diff
@@ -739,6 +739,22 @@ impl<'tcx> ty::TyS<'tcx> {
         tcx_at.is_conservative_raw(param_env.and(self))
     }
 
+    pub fn needs_tracing(
+        &'tcx self,
+        tcx_at: TyCtxtAt<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+    ) -> bool {
+        self.is_conservative(tcx_at, param_env) || !self.is_no_trace(tcx_at, param_env)
```

Comment:
Should `||` be `&&`?

Reply:
No, `is_conservative` is like an override trait that takes precedence over `NoTrace`.

Reply:
Dumb question: why would someone need to do such an override?

Reply:
`NoTrace` implies `alloc_untraceable`.

`!NoTrace` implies `alloc_precise` _or_ `alloc_conservative` but never `alloc_untraceable`.

Until we had `trait Conservative`, there was no single trait that could force the optimiser to always choose the safe-but-slow option (`alloc_conservative`). This is necessary for types like `SOMStack` and also blocks of memory where `layout` doesn't directly correspond to a `T`.

Reply:
Aha, I think I get it!